### ISSUE TITLE
AutoValue incorrectly resolves generics in Builder superinterfaces

### ISSUE THREAD

**Issue by ekursakov** [google/auto#627]
I have two interfaces:
```java
public interface ItemsHolder<T> {
    List<T> getItems();
}
public interface ItemsHolderBuilder<Self extends ItemsHolderBuilder<Self, T>, T> {
    Self setItems(List<T> items);
}
```
When I am trying to implement them in AutoValue class like this
```java
@AutoValue
public abstract class GroceryList implements ItemsHolder<GroceryItem> {
    @AutoValue.Builder
    public abstract static class Builder implements ItemsHolderBuilder<Builder, GroceryItem> {
        public abstract GroceryList build();
    }
}
```
AutoValue generates incorrect code for setter:
```java
final class AutoValue_GroceryList extends GroceryList {
    private final List<GroceryItem> items;

    private AutoValue_GroceryList(List<GroceryItem> items) {
        this.items = items;
    }

    @Override
    public List<GroceryItem> getItems() {
        return items;
    }
    ...
    static final class Builder extends GroceryList.Builder {
        private List<GroceryItem> items;

        Builder() { }

        @Override
        public GroceryList.Builder setItems(List<T> items) { // <-- error: cannot find symbol class T
            if (items == null) {
                throw new NullPointerException("Null items");
            }
            this.items = items;
            return this;
        }

        @Override
        public GroceryList build() {
            ...
        }
    }
}
```
As a workaround I had to [override these methods](https://github.com/ekursakov/AutoValueTest/blob/master/app/src/main/java/ek/autovaluetest/GroceryList.java) in `GroceryList` and builder.
Sample project: https://github.com/ekursakov/AutoValueTest 
AutoValue version is 1.6

**Comment by ekursakov** [comment]
Found PR #515 that should resolve this

**Comment by eamonnmcmanus** [comment]
I just commented on that PR. The issue is that that PR uses [Types.asMemberOf](https://docs.oracle.com/javase/8/docs/api/javax/lang/model/util/Types.html#asMemberOf-javax.lang.model.type.DeclaredType-javax.lang.model.element.Element-) but the Eclipse compiler has historically failed to implement that method correctly, throwing exceptions in all but the simplest cases. It might be that that bug has been fixed, in which case this and certain other fixes would be possible.
